docs/expiry-and-auto-renewal/renewal_expiry_design_v2.md

```diff
@@ -0,0 +1,279 @@
+architecture behind AutoRenewal of accounts and its capabilities to be used for tokens (fungible and non-fungible)
+expiry and auto renewal. We're considering the following logical framework in order to support a more generic expiration
+handling, with those **main** elements:
+
+-------------
+
+### Worker
+
+The worker is Hedera's vision of a background process - a process which is executed often enough. The current
+expiries/renewals for accounts are triggered on each transaction. The background worker is responsible for executing all
+actions during the post-transaction phase. This is possible because of the bucket abstraction implied.
+
+Below is `Worker` functional interface:
+
+```java
+interface Worker {
+	boolean run(long consensusTimeStamp);
+}
+```
+
+This method implementation will iterate over all buckets and execute the determined actions. The worker will also pass
+the available network capacity through the Throttling system to each bucket.
+
+-------------
+
+### Bucket
+
+The bucket is a logical group of executable actions, grouped by common functionality. A bucket's responsibility may be
+account scanner and deletion. New buckets can be introduced in future for token scanning and removal.
+
+For migrating the current expiry/renewal account implementation, we might consider 2 types of buckets
+- `AccountScannerBucket` and `AccountDeletionBucket`.
+
+Below is `Bucket` functional interface:
+
+```java
+interface Bucket {
+	/* Returns the capacity used by this work. */
+	Capacity drain(long consensusTimeStamp, Capacity available);
+}
+```
+
+The implementation `AccountScannerBucket` will be responsible for:
+
+- iteration over all merkle accounts
+- for every iteration (entity touch) the capacity will be depleted
+- determination of action for each account loaded from state. The returned actions must be only for scanning purposes.
+  For instance if account must be renewed, `Renew` action will be returned
+- action execution
+
+Pseudo code of the drain method:
+
+````java
+class AccountScannerBucket {
+	public Capacity drain(long consensusTimeStamp, Capacity available) {
+		while (hasEnoughCapacity(available)) {
+			MerkleAccount account = loadAccount(++lastScannedEntity);
+			Option<Action> action = determineAction(account, now);
+
+			if (!action.isEmpty() && hasAvailableCapacity(availability, action.requiredCapacity())) {
+              			action.execute();
+              			availability -= action.requiredCapacity() + action.requiredCapacity() * throttleLoad();
+              			lastScannedEntity = accountNum;
+			}
+		}
+
+		return available;
+	}
+}
+````
+
+The implementation of `AccountDeletionBucket` will be responsible for:
+
+- iteration over all account relations
+- for every iteration (entity touch) the capacity will be depleted
+- determination of action for each merkle account loaded from state
+- action execution
+- removal of account from `MerkleNetworkContext` queue
```

Comment:
- This is valid if the `action` is `AccountRemoval`
- This step is not added in the pseudo code. If it is moved in the `Action`, please remove it from the description of the bucket